### ISSUE TITLE
ValidatePattern should be avoided in some (many?) cases

### ISSUE THREAD

**Issue by nightroman** [darkoperator/PSStyleGuide#18]
> avoid validating parameters in the body of the script when possible and use
parameter validation attributes instead.

> The ValidatePattern attribute specifies a regular expression that is compared
to the parameter or variable value. Windows PowerShell generates an error if
the value does not match the regular expression pattern.

From my experience `ValidatePattern` is a feature to be avoided when possible,
in contrast to the guidelines. It's regular expressions and end users are not
necessarily experts. But PowerShell will show them just regular expressions in
error messages, nothing else. Even simple expressions like `^\w+$` are cryptic
for some people. Compare with "should contain alphanumeric and underscore
characters."

P.S. I agree with other parameter attributes, they are good. `ValidateScript`
also has some issues (see [here]) but unlike with `ValidatePattern` they can be
resolved with using `throw <friendly message>`.

[here]: https://github.com/nightroman/PowerShellTraps/tree/master/Basic/ValidateScript-attribute

**Comment by Jaykul** [comment]
I agree. When patterns are needed, I recommend always using ValidateScript with a -match and a throw that describes the problem in English. I'm not sure we really need to discuss all the validation attributes anyway, but that one in particular I *would* mention (and recommend against).

@darkoperator Is there a reason you repeated so much of [about_Functions_Advanced_Parameters](https://technet.microsoft.com/en-us/library/hh847743.aspx) in there instead of just referring to it for the validation attributes?

**Comment by darkoperator** [comment]
I repeated it just for consistency, I agree that one should be modified

**Comment by pixelrebirth** [comment]
Why wouldnt you just try the function outside the function block and convert to human readable there? This is how I do it and all error handling is centralized script wide!

ex.

function Test-RegexString {
    param (
        [ValidatePattern('^\w+$')]
        [string]
        $inputString
    )
    $inputString
}


function Get-ErrorHandle {    
    if ($error[0] -like "*^\w+$*") { 
        write-error "The -inputSting should contain alphanumeric and underscore characters." 
    }
}


try { Test-RegexString }
catch { Get-ErrorHandle }

**Comment by Jaykul** [comment]
I'm not sure I understand your design @pixelrebirth -- that won't help people who are calling your Test-RegexString.  I think what you should be doing to centralize error handling and messages is something like this:

```posh
function ShouldBeAlphaNumeric {
param($name, $string)
    if ($string -notmatch "^\w+$") { 
        throw "The $name should only contain alphanumeric and underscore characters." 
    }
    $true
}


function Test-RegexString {
    param (
        [ValidateScript({ShouldBeAlphaNumeric inputstring $_})]
        [string]
        $inputString
    )
    $inputString
}
```

That way the `Test-RegexString` function has proper errors regardless of who calls it and how.

**Comment by pixelrebirth** [comment]
I see that perspective, particular for public functions that makes sense. I guess what I was trying to accomplish is a master error handler:

```
function Test-RegexString {
    param (
        [ValidatePattern('^\w+$')]
        [string]
        $inputString
    )
    $inputString
}


function Get-ErrorHandle {    
    if ($error[0] -like "*^\w+$*") { 
        write-error "The input should contain alphanumeric and underscore characters." 
    }
    if ($error[0] -like "*^d$") { 
        write-error "The input should be a number." 
    }
    if ( $error[0] -like "*some wack third party error message*" ) {
        write-error "You need a new piece of software."
    }
}


try { Test-RegexString "!" }
catch { Get-ErrorHandle }

**Comment by Jaykul** [comment]
I don't have any problem with the idea of a central error handling function @pixelrebirth (I actually like the idea of dealing with different common exception types that way)

However, it doesn't address the problem -- which is that you should never have used ValidatePattern in the first place!  It's a horrible user experience, and leads people to having to write ... well ... try catch blocks to produce better error messages :laughing:

**Comment by pixelrebirth** [comment]
I had a discussion with a colleague and I am on the same boat as you now. Create a file exclusive for the error functions and exclusively use validatescript (functionname) --- you can even aggregate the functions into a single get-errorhandle function for readability without having to open another file to see which function it matches. IE.

```powershell
function Get-ErrorHandle {  
    ShouldBeAlphaNumeric #imported function
    ShouldBeDigit #imported function
    ShouldBeBetterSoftware #imported function
}

function Test-RegexString {
    param (
        [ValidateScript({ShouldBeAlphaNumeric inputstring $_})]
        [string]
        $inputString
    )
    $inputString
}

try { Test-RegexString "!" }
catch { Get-ErrorHandle }
```

**Comment by pixelrebirth** [comment]
I wonder if the exclusive use of ValidateScript over any other validation class makes sense? even for something easily readable like ValidateLength or ValidateNotNullOrEmpty --- could be more human readable with custom error functions.

**Comment by Jaykul** [comment]
Well, ValidateSet has a side benefit of being picked up by tab expansion and intellisense, so I definitely recommend using that when appropriate.

As for the others, I think the same thing is true: editors and graphical tools can use the fact that they are well known common validators to improve their intellisense and prompting. 

If you think the message could be better, you should file a bug, rather than just using ValidateScript -- it's better to have a uniform (and presumably localized?) error message (that is, if you use the built-in validators, you get a good error message that people are used to seeing because everyone else uses them too).

Speaking of that, vote for this https://connect.microsoft.com/PowerShell/feedbackdetail/view/1450924/the-validatepattern-attribute-should-have-a-failure-message-optional-parameter
